File: source/java/src/main/java/cmf/eventing/berico/DigitalSignatureProcessor.java
package cmf.eventing.berico;

import java.security.Signature;
import java.security.cert.X509Certificate;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import cmf.bus.berico.EnvelopeHelper;
import cmf.security.CredentialHolder;
import cmf.security.ICertificateProvider;

public class DigitalSignatureProcessor implements IInboundEventProcessor, IOutboundEventProcessor {

    protected ICertificateProvider certProvider;
    protected CredentialHolder credentials;
    protected Logger log;

    public DigitalSignatureProcessor(ICertificateProvider certProvider) throws Exception {
        this.certProvider = certProvider;
        log = LoggerFactory.getLogger(this.getClass());

        try {
            credentials = certProvider.getCredentials();
        } catch (Exception ex) {
            log.error("Failed to get client's PKI credentials", ex);
            throw ex;
        }
    }

    @Override
    public boolean processInbound(ProcessingContext context) throws Exception {

        boolean verified = false;

        try {
            EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());
            String senderIdentity = env.getSenderIdentity();

            // remove spaces from the sender identity
            senderIdentity = senderIdentity.replace(", ", ",");

            X509Certificate senderCert = certProvider.getCertificateFor(senderIdentity);

            Signature instance = Signature.getInstance("SHA1withRSA");
            instance.initVerify(senderCert);
            instance.update(env.getPayload());

            verified = instance.verify(env.getDigitalSignature());

            if (false == verified) {
                log.warn("Event may have been tampered with (id:{})", env.getMessageId().toString());
            }
        } catch (Exception ex) {
            log.error("Failed to verify sender's digital signature", ex);
            throw ex;
        }

        return verified;
    }

    @Override
    public void processOutbound(ProcessingContext context) throws Exception {

        try {
            EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());

            Signature instance = Signature.getInstance("SHA1withRSA");
            instance.initSign(credentials.getPrivateKey());
            instance.update(env.getPayload());

            env.setDigitalSignature(instance.sign());

            // make sure that the sender identity has spaces between LDAP elements
            String sender = credentials.getCertificate().getSubjectX500Principal().getName();
            sender = sender.replaceAll(",", ", ");

            env.setSenderIdentity(sender);
        } catch (Exception ex) {
            log.error("Exception while signing outbound event", ex);
            throw ex;
        }
    }
}

File: dotnet/source/cmf.eventing.berico/DigitalSignatureProcessor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Text;

using Common.Logging;

using cmf.bus;
using cmf.bus.berico;
using cmf.security;

namespace cmf.eventing.berico
{
    public class DigitalSignatureProcessor : IInboundEventProcessor, IOutboundEventProcessor
    {
        protected ICertificateProvider _certProvider;
        protected X509Certificate2 _cert;
        protected ILog _log;


        public DigitalSignatureProcessor(ICertificateProvider certProvider)
        {
            _certProvider = certProvider;

            try
            {
                _cert = certProvider.GetCertificate();
            }
            catch (Exception ex)
            {
                _log.Error("Failed to get a certificate from the provider", ex);
            }
        }


        public bool ProcessInbound(ref object ev, ref Envelope env, ref IDictionary<string, object> context)
        {
            bool signatureVerified = false;

            // start by getting the purported sender's identity
            string sender = env.GetSenderIdentity();
            if (string.IsNullOrEmpty(sender))
            {
                _log.Error("An inbound event arrived with no sender identity");
            }
            else
            {
                // use the identity to lookup their certificate
                X509Certificate2 senderCert = _certProvider.GetCertificateFor(sender);
                if (null == senderCert)
                {
                    _log.Error("Sender " + sender + " does not have a public key certificate available");
                }
                else
                {
                    // get the digital signature from the headers
                    byte[] digitalSignature = env.GetDigitalSignature();
                    if (0 == digitalSignature.LongLength)
                    {
                        _log.Error("Sender " + sender + " did not digitally sign the event");
                    }
                    else
                    {
                        // verify that the payload hasn't been tampered with by using the 
                        // sender's public key and the digital signature on the envelope
                        RSACryptoServiceProvider rsaProvider = senderCert.PublicKey.Key as RSACryptoServiceProvider;
                        signatureVerified = rsaProvider.VerifyData(env.Payload, new SHA1CryptoServiceProvider(), digitalSignature);
                    }
                }
            }

            return signatureVerified;
        }

        public void ProcessOutbound(ref object ev, ref Envelope env, IDictionary<string, object> context)
        {
            if (null == _cert) { return; }

            env.SetSenderIdentity(_cert.Subject);

            try
            {
                RSACryptoServiceProvider rsaProvider = _cert.PrivateKey as RSACryptoServiceProvider;
                env.SetDigitalSignature(rsaProvider.SignData(env.Payload, new SHA1CryptoServiceProvider()));
            }
            catch (Exception ex)
            {
                _log.Error("Failed to digitally sign the event", ex);
                throw;
            }
        }
    }
}
